File: src/services/reasoningCatalog.ts
import catalog from '../shared/ai/reasoningCatalog.json';
import type {
  ReasoningCapability,
  ReasoningCapabilitySource,
  ReasoningEffort,
  ReasoningTransportMode,
} from '../types';
import { lookupModelReasoningCatalogCapability } from './modelContextCatalog';

interface ReasoningCatalogEntry {
  id: string;
  provider_types: string[];
  model_patterns: string[];
  supported_efforts: string[];
  default_effort: string;
  configurable?: boolean;
  transport_mode?: ReasoningTransportMode;
}

interface ReasoningCatalogFile {
  version: number;
  entries: ReasoningCatalogEntry[];
}

const reasoningCatalog = catalog as ReasoningCatalogFile;

export const CANONICAL_REASONING_EFFORTS = [
  'none',
  'minimal',
  'low',
  'medium',
  'high',
  'xhigh',
  'max',
] as const satisfies readonly ReasoningEffort[];

const CANONICAL_EFFORT_ORDER = new Map<string, number>(
  CANONICAL_REASONING_EFFORTS.map((effort, index) => [effort, index]),
);

const SAFE_REASONING_EFFORT = /^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$/;

export const normalizeReasoningEffortValue = (value: unknown): ReasoningEffort | null =>
  typeof value === 'string' && SAFE_REASONING_EFFORT.test(value) ? value : null;

const EMPTY_REASONING_CAPABILITY: ReasoningCapability = {
  reasoningEfforts: [],
  defaultReasoningEffort: null,
  transportMode: 'none',
  configurable: false,
  source: 'none',
};

export const normalizeReasoningEfforts = (
  value?: readonly string[] | null,
): ReasoningEffort[] => {
  if (!Array.isArray(value)) return [];
  const seen = new Set<string>();
  const normalized = value
    .map((entry, inputIndex) => ({ effort: normalizeReasoningEffortValue(entry), inputIndex }))
    .filter((entry): entry is { effort: ReasoningEffort; inputIndex: number } => {
      if (entry.effort === null || seen.has(entry.effort)) return false;
      seen.add(entry.effort);
      return true;
    });

  return normalized
    .sort((left, right) => {
      const leftOrder = CANONICAL_EFFORT_ORDER.get(left.effort);
      const rightOrder = CANONICAL_EFFORT_ORDER.get(right.effort);
      if (leftOrder === undefined && rightOrder === undefined) {
        return left.inputIndex - right.inputIndex;
      }
      if (leftOrder === undefined) return 1;
      if (rightOrder === undefined) return -1;
      return leftOrder - rightOrder;
    })
    .map(({ effort }) => effort);
};

const firstValidEffort = (
  preferred: string | null | undefined,
  efforts: readonly ReasoningEffort[],
): ReasoningEffort | null => {
  const normalizedPreferred = preferred?.trim();
  if (normalizedPreferred && efforts.includes(normalizedPreferred)) return normalizedPreferred;
  return efforts[0] ?? null;
};

const hasOpenRouterReasoningSupport = (supportedParameters?: readonly string[] | null): boolean => {
  if (!Array.isArray(supportedParameters)) return false;
  return supportedParameters.includes('reasoning') || supportedParameters.includes('reasoning_effort');
};

const matchesCatalogEntry = (
  providerType: string,
  modelId: string,
  entry: ReasoningCatalogEntry,
): boolean =>
  entry.provider_types.includes(providerType) &&
  entry.model_patterns.some((pattern) => new RegExp(pattern, 'i').test(modelId));

const resolveTransportMode = (
  providerType: string,
  modelId: string,
  configuredMode?: ReasoningTransportMode,
): ReasoningTransportMode => {
  if (providerType === 'openrouter') return 'openrouter_reasoning';
  if (configuredMode) return configuredMode;
  if (/deepseek/i.test(modelId)) return 'deepseek_thinking';
  if (/kimi|moonshot/i.test(modelId)) return 'kimi_fixed';
  return 'openai_effort';
};

const buildCapability = (params: {
  efforts: readonly string[];
  preferredDefault?: string | null;
  providerType: string;
  modelId: string;
  source: ReasoningCapabilitySource;
  configurable?: boolean;
  transportMode?: ReasoningTransportMode;
}): ReasoningCapability => {
  const reasoningEfforts = normalizeReasoningEfforts(params.efforts);
  if (reasoningEfforts.length === 0) return EMPTY_REASONING_CAPABILITY;
  return {
    reasoningEfforts,
    defaultReasoningEffort: firstValidEffort(params.preferredDefault, reasoningEfforts),
    transportMode: resolveTransportMode(params.providerType, params.modelId, params.transportMode),
    configurable: params.configurable ?? reasoningEfforts.length > 1,
    source: params.source,
  };
};

export const getReasoningCapabilityForModel = (params: {
  providerType?: string | null;
  providerId?: string | null;
  baseUrl?: string | null;
  modelId?: string | null;
  supportedParameters?: readonly string[] | null;
  supportedReasoningEfforts?: readonly string[] | null;
  defaultReasoningEffort?: string | null;
  manualReasoningEfforts?: readonly string[] | null;
  manualDefaultReasoningEffort?: string | null;
}): ReasoningCapability => {
  const providerType = params.providerType?.trim().toLowerCase();
  const modelId = params.modelId?.trim();
  if (!providerType || !modelId) return EMPTY_REASONING_CAPABILITY;

  const manualEfforts = normalizeReasoningEfforts(params.manualReasoningEfforts);
  if (manualEfforts.length > 0) {
    return buildCapability({
      efforts: manualEfforts,
      preferredDefault: params.manualDefaultReasoningEffort,
      providerType,
      modelId,
      source: 'manual_override',
      configurable: true,
    });
  }

  const directEfforts = normalizeReasoningEfforts(params.supportedReasoningEfforts);
  if (directEfforts.length > 0) {
    return buildCapability({
      efforts: directEfforts,
      preferredDefault: params.defaultReasoningEffort,
      providerType,
      modelId,
      source: 'provider_metadata',
    });
  }

  if (providerType === 'openrouter' && !hasOpenRouterReasoningSupport(params.supportedParameters)) {
    return EMPTY_REASONING_CAPABILITY;
  }

  const modelsDev = lookupModelReasoningCatalogCapability({
    providerType,
    providerId: params.providerId,
    baseUrl: params.baseUrl,
    modelId,
  });
  if (modelsDev && modelsDev.reasoningEfforts.length > 0) {
    return buildCapability({
      efforts: modelsDev.reasoningEfforts,
      preferredDefault: modelsDev.defaultReasoningEffort,
      providerType,
      modelId,
      source: 'models_dev',
      configurable: modelsDev.configurable,
      transportMode: modelsDev.transportMode,
    });
  }

  const matched = reasoningCatalog.entries.find((entry) =>
    matchesCatalogEntry(providerType, modelId, entry),
  );
  if (!matched) return EMPTY_REASONING_CAPABILITY;
  return buildCapability({
    efforts: matched.supported_efforts,
    preferredDefault: matched.default_effort,
    providerType,
    modelId,
    source: 'embedded_catalog',
    configurable: matched.configurable,
    transportMode: matched.transport_mode,
  });
};

export const getValidReasoningEffort = (
  capability: Pick<ReasoningCapability, 'reasoningEfforts' | 'defaultReasoningEffort'>,
  requested?: string | null,
): ReasoningEffort | null => {
  if (requested && capability.reasoningEfforts.includes(requested)) return requested;
  return capability.defaultReasoningEffort ?? capability.reasoningEfforts[0] ?? null;
};

export const getReasoningLabel = (effort: ReasoningEffort): string => {
  switch (effort) {
    case 'none': return 'None';
    case 'minimal': return 'Minimal';
    case 'low': return 'Low';
    case 'medium': return 'Medium';
    case 'high': return 'High';
    case 'xhigh': return 'X-High';
    case 'max': return 'Max';
    default: return effort;
  }
};

File: src-tauri/src/ai/reasoning_catalog.rs
use regex::Regex;
use serde::Deserialize;
use std::collections::HashSet;
use std::sync::LazyLock;

#[derive(Debug, Clone, Deserialize)]
struct ReasoningCatalogFile {
    #[allow(dead_code)]
    version: u32,
    entries: Vec<ReasoningCatalogEntry>,
}

#[derive(Debug, Clone, Deserialize)]
struct ReasoningCatalogEntry {
    provider_types: Vec<String>,
    model_patterns: Vec<String>,
    supported_efforts: Vec<String>,
    default_effort: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ReasoningCapability {
    pub reasoning_efforts: Vec<String>,
    pub default_reasoning_effort: Option<String>,
}

static REASONING_CATALOG: LazyLock<ReasoningCatalogFile> = LazyLock::new(|| {
    serde_json::from_str(include_str!("../../../src/shared/ai/reasoningCatalog.json"))
        .expect("reasoning catalog must be valid JSON")
});

static CANONICAL_EFFORTS: [&str; 7] = ["none", "minimal", "low", "medium", "high", "xhigh", "max"];

fn is_safe_effort(effort: &str) -> bool {
    let mut chars = effort.chars();
    let Some(first) = chars.next() else {
        return false;
    };
    effort.len() <= 64
        && first.is_ascii_alphanumeric()
        && chars.all(|character| {
            character.is_ascii_alphanumeric() || matches!(character, '.' | '_' | '-')
        })
}

fn normalize_efforts(efforts: Option<&[String]>) -> Vec<String> {
    let mut seen = HashSet::new();
    let mut normalized = efforts
        .into_iter()
        .flatten()
        .filter(|effort| is_safe_effort(effort) && seen.insert((*effort).clone()))
        .cloned()
        .enumerate()
        .collect::<Vec<_>>();
    normalized.sort_by_key(|(input_index, effort)| {
        (
            CANONICAL_EFFORTS
                .iter()
                .position(|known| known == effort)
                .unwrap_or(CANONICAL_EFFORTS.len()),
            *input_index,
        )
    });
    normalized.into_iter().map(|(_, effort)| effort).collect()
}

#[cfg(test)]
mod tests {
    use super::{normalize_efforts, resolve_reasoning_capability};

    #[test]
    fn preserves_max_and_safe_provider_defined_efforts() {
        let efforts = vec![
            "provider_custom".to_string(),
            "max".to_string(),
            "low".to_string(),
            "invalid effort".to_string(),
        ];
        assert_eq!(
            normalize_efforts(Some(&efforts)),
            vec!["low", "max", "provider_custom"]
        );
    }

    #[test]
    fn resolves_current_gpt_5_6_catalog_entry() {
        let capability =
            resolve_reasoning_capability(Some("openai"), Some("gpt-5.6"), None, None, None);
        assert!(capability.reasoning_efforts.contains(&"max".to_string()));
        assert_eq!(
            capability.default_reasoning_effort.as_deref(),
            Some("medium")
        );
    }
}

fn first_valid_effort(preferred: Option<&str>, efforts: &[String]) -> Option<String> {
    if let Some(preferred) = preferred {
        if efforts.iter().any(|effort| effort == preferred) {
            return Some(preferred.to_string());
        }
    }

    efforts.first().cloned()
}

fn has_openrouter_reasoning_support(supported_parameters: Option<&[String]>) -> bool {
    supported_parameters
        .map(|parameters| {
            parameters
                .iter()
                .any(|parameter| parameter == "reasoning" || parameter == "reasoning_effort")
        })
        .unwrap_or(false)
}

pub fn resolve_reasoning_capability(
    provider_type: Option<&str>,
    model_id: Option<&str>,
    supported_parameters: Option<&[String]>,
    supported_reasoning_efforts: Option<&[String]>,
    default_reasoning_effort: Option<&str>,
) -> ReasoningCapability {
    let Some(provider_type) = provider_type
        .map(str::trim)
        .filter(|value| !value.is_empty())
    else {
        return ReasoningCapability {
            reasoning_efforts: Vec::new(),
            default_reasoning_effort: None,
        };
    };
    let Some(model_id) = model_id.map(str::trim).filter(|value| !value.is_empty()) else {
        return ReasoningCapability {
            reasoning_efforts: Vec::new(),
            default_reasoning_effort: None,
        };
    };

    let direct_efforts = normalize_efforts(supported_reasoning_efforts);
    if !direct_efforts.is_empty() {
        return ReasoningCapability {
            default_reasoning_effort: first_valid_effort(default_reasoning_effort, &direct_efforts),
            reasoning_efforts: direct_efforts,
        };
    }

    if provider_type.eq_ignore_ascii_case("openrouter")
        && !has_openrouter_reasoning_support(supported_parameters)
    {
        return ReasoningCapability {
            reasoning_efforts: Vec::new(),
            default_reasoning_effort: None,
        };
    }

    let matched = REASONING_CATALOG.entries.iter().find(|entry| {
        entry
            .provider_types
            .iter()
            .any(|supported_provider| supported_provider.eq_ignore_ascii_case(provider_type))
            && entry.model_patterns.iter().any(|pattern| {
                Regex::new(pattern)
                    .map(|compiled| compiled.is_match(model_id))
                    .unwrap_or(false)
            })
    });

    let Some(matched) = matched else {
        return ReasoningCapability {
            reasoning_efforts: Vec::new(),
            default_reasoning_effort: None,
        };
    };

    let reasoning_efforts = normalize_efforts(Some(&matched.supported_efforts));
    ReasoningCapability {
        default_reasoning_effort: first_valid_effort(
            Some(&matched.default_effort),
            &reasoning_efforts,
        ),
        reasoning_efforts,
    }
}
